### PR TITLE
Adds a new option to automagically calculate memory from available memory on system

### PR DESCRIPTION
**Pull request by standage**
See #1249. This adds a new dependency [psutil](https://github.com/giampaolo/psutil), which is available via pip and actively developed.

---
- [x] Is it mergeable?
- [x] `make test` Did it pass the tests?
- [x] `make clean diff-cover` If it introduces new functionality in
  `scripts/` is it tested?
- [x] `make format diff_pylint_report cppcheck doc pydocstyle` Is it well
  formatted?
- [x] Did it change the command-line interface? Only additions are allowed
  without a major version increment. Changing file formats also requires a
  major version number increment.
- [x] Is it documented in the `ChangeLog`?
  http://en.wikipedia.org/wiki/Changelog#Format
- [x] Was a spellchecker run on the source code and documentation after
  changes were made?
- [x] Do the changes respect streaming IO? (Are they
  tested for streaming IO?)
- [x] Is the Copyright year up to date?

### CODE REVIEW

File: khmer/khmer_args.py [hunk 1]
```diff
@@ -356,25 +361,29 @@ def add_loadgraph_args(parser):
                         help='load a precomputed k-mer graph from disk')
 
 
-def calculate_graphsize(args, graphtype, multiplier=1.0):
+def calculate_graphsize(args, graphtype, multiplier=1.0,
+                        sysmem=psutil.virtual_memory().available):
```
Comment (ctb) [inline khmer/khmer_args.py]: I'd suggest leaving this as something default ('None'?) and then calculating it in body of the function if needed; then you can pass it in for the tests w/o making the default def that ugly.

Also, if it's an internal thing that will only be used by tests, suggest naming it _sysmem or something else.

Comment (standage) [inline khmer/khmer_args.py]: Yeah, that's a bit cleaner. And I can also avoid cluttering the `create_*graph` functions by testing `calculate_graphsize` directly. That'll also help avoid problems with Jenkins like when I, oh, tried to create a 64 GB countgraph for testing! :-/

**Comment by standage** [comment]
Ready for review and merge @ctb et al.

**Comment by ctb** [comment]
Hi @standage, thx -

A few questions/comments off the top:

* is psutil a binary dependency, or is it pure Python?
* I think your first comment is no longer true :)
* in #1249, we suggest having auto-mem be the default behavior; that seems OK to me, as long as we're explicit about it. ...but this might require a major version number. Hrm.

**Comment by ctb** [comment]
(Note that grabbing a pile of memory is what most programs do. So I don't think we should feel guilty about it, since at least ours won't crash the system, probably.)

**Comment by standage** [comment]
>  is psutil a binary dependency, or is it pure Python?

It appears psutil has some compiled C code.

> I think your first comment is no longer true :)

Correct. I thought I added a clarifying remark somewhere, but I don't see it on this thread.

> in #1249, we suggest having auto-mem be the default behavior; that seems OK to me, as long as we're explicit about it. ...but this might require a major version number. Hrm.

Yeah, making this the default would indeed be grounds for a major version bump. But what about adding the option with v2.1, and then waiting to make it the default until v3.0 release?

**Comment by standage** [comment]
I've done a bit more searching, and most of the solutions I see for determining available system memory involve parsing `/proc/meminfo`, which is a Linux-specific solution. A pure Python package with this type of cross-platform OS awareness doesn't seem readily available, and (in my opinion) is beyond the scope of what we'd want to implement within this project.

**Comment by ctb** [comment]
On Tue, Jul 05, 2016 at 09:16:17AM -0700, Daniel Standage wrote:
> >  is psutil a binary dependency, or is it pure Python?
> 
> It appears psutil has some compiled C code.
> 
> > I think your first comment is no longer true :)
> 
> Correct. I thought I added a clarifying remark somewhere, but I don't see it on this thread.

OK. I tend to edit the head comment matter for this kind of thing, YMMV.

> > in #1249, we suggest having auto-mem be the default behavior; that seems OK to me, as long as we're explicit about it. ...but this might require a major version number. Hrm.
> 
> Yeah, making this the default would indeed be grounds for a major version bump. But what about adding the option with v2.1, and then waiting to make it the default until v3.0 release?

:+1:

File: khmer/khmer_args.py [hunk 2]
```diff
@@ -360,12 +361,15 @@ def build_graph_args(descr=None, epilog=None, parser=None):
     group = parser.add_mutually_exclusive_group()
     group.add_argument('--max-tablesize', '-x', type=float,
                        default=DEFAULT_MAX_TABLESIZE,
-                       help='upper bound on tablesize to use; overrides ' +
+                       help='upper bound on tablesize to use; overrides '
                        '--max-memory-usage/-M')
```
Comment (betatim) [inline khmer/khmer_args.py]: One thing you can do is add `(` `)` around the multi line string and indent the second line to line up with the one above. I find it helps with making it clear that these strings were meant to be joined together. Not sure what the style gods say regarding implicit/explicit joining of strings across lines.

File: khmer/khmer_args.py [hunk 3]
```diff
@@ -360,12 +361,15 @@ def build_graph_args(descr=None, epilog=None, parser=None):
...
-    group.add_argument('-M', '--max-memory-usage', type=memory_setting,
-                       help='maximum amount of memory to use for data ' +
+    group.add_argument('-M', '--max-memory-usage', type=float,
+                       help='maximum amount of memory to use for data '
                        'structure')
```
Comment (betatim) [inline khmer/khmer_args.py]: same as above

File: khmer/khmer_args.py [hunk 4]
```diff
@@ -360,12 +361,15 @@ def build_graph_args(descr=None, epilog=None, parser=None):
...
-
+    group.add_argument('-a', '--auto-mem', action='store_true',
+                       help='automagically allocate memory for data structure;'
+                       ' use 80%% of available memory or 16GB, whichever is '
+                       'smaller')
```
Comment (betatim) [inline khmer/khmer_args.py]: same as above

File: khmer/khmer_args.py [hunk 5]
```diff
@@ -389,18 +393,22 @@ def add_loadgraph_args(parser):
                         help='load a precomputed k-mer graph from disk')
 
 
-def calculate_graphsize(args, graphtype, multiplier=1.0):
+def calculate_graphsize(args, graphtype, multiplier=1.0, _sysmem=None):
```
Comment (betatim) [inline khmer/khmer_args.py]: Why have the `_sysmem` argument? Is it solely to make the unittests work? In python you can use [mock](https://docs.python.org/3/library/unittest.mock.html) to replace the `psutil` call for the unit test. Unfortunately it isn't part of the stdlib for `python2`.

(Function arguments starting with a `_` attract attention in python land 👀

File: khmer/khmer_args.py (hunk 2 above)
Comment (ctb) [inline khmer/khmer_args.py]: 'make pep8' checks against our chosen style standard; if it passes that,
we're happy :)

File: khmer/khmer_args.py (hunk 5 above)
Comment (standage) [inline khmer/khmer_args.py]: Yes, it's solely for the unit tests.

By "attract attention", do you mean *bad* attention? I thought the leading underscore was a pretty common convention for saying "this is internal, you shouldn't mess with this unless you know what you're doing."

Comment (betatim) [inline khmer/khmer_args.py]: Attracts attention in the sense that I'd have to think about it before I could show you another example of it. `_` in python does indeed mean "don't mess" though usually apply to member variables or functions/methods.

TL;DR: There should be a better way of doing this (but I can't think of one right now short of introducing `mock` as a dependency for `python2`). Maybe an executive decision is needed re adding dependencies for developers.

Comment (ctb) [inline khmer/khmer_args.py]: If it causes practical problems, we'll eliminate it.  Hard to know at this
point. But memory settings are one of the major causes of user pain,
at least anecdotally.

Comment (betatim) [inline khmer/khmer_args.py]: The downside to having the argument there is that one day a user will try and use it. I don't think it will cause technical problems. The pythonista in me thinks it is ugly, in the end it is like all code smells...

Comment (ctb) [inline khmer/khmer_args.py]: Sure; the tradeoff is that the workaround would add technical complexity
to our testing framework :)

**Comment by standage** [comment]
Reflecting on this several months later: we have minor syntax/implementation issues that IMO aren't blockers for merging. If we still feel strongly about them we can quickly fix, if not we can open separate issues to continue (not) discussing.

The big question here is whether we want to add psutil as a dependency. It's not pure Python, so it will complication installation, especially on legacy systems. The counter argument is that now we're requiring C++11 support anyway, and since this is a robust and actively developed library it should compile on the vast majority of "modern" systems.

What I DON'T think we should do is try to implement OS-dependent memory awareness ourselves. That'll be messy and buggy.

Honestly, now that khmer supports human-friendly memory requests (`-M 8G`), much of the user headache should be gone. I'm in favor of closing this PR for now, or punting.

**Comment by ctb** [comment]
punt with #wontfix then. thx!
